### DISCUSSION TITLE
The cron job problem.

### DISCUSSION THREAD

**Discussion by Kgitman** [supabase/supabase#29265] [Questions] [answered]
hello, everyone! @GaryAustin1 and team,

I have the `edge function` where I fetching the data and than storing that data into the supabase. now I want to schedule the edge function like with cron or something so it can be invoke at a prefered time.

I use this format before but I don't want to use now because I'm `exponsing` the `anon key` in various oprations and it's neccessary.

```
select
  cron.schedule(
    'invoke-function-every-minute',
    '* * * * *', -- every minute
    $$
    select
      net.http_post(
          url:='https://project-ref.supabase.co/functions/v1/function-name',
          headers:='{"Content-Type": "application/json", "Authorization": "Bearer YOUR_ANON_KEY"}'::jsonb,
          body:=concat('{"time": "', now(), '"}')::jsonb
      ) as request_id;
    $$
  );
```

Now I enabled the RLS and selected the `service role key` to make my table secure and all oprations like insert and update with `service_role_key` because I'm exposing the `anon key`!

But I'm not understanding how to run the cron job now? 

If I paste here the `service_role_key` in place of `anon_key `
`"Authorization": "Bearer YOUR_ANON_KEY"`

than the cron job also failed may be because I'm passing the `anon_key` in the edge function.

![image](https://github.com/user-attachments/assets/6e0bd2bc-ecd3-4bec-b595-f7a2d1430b98)

And if I pass the service role key than the edge function will invoke with both the `anon_key` and the `service_role_key` which is also a threat that anyone can invoke the edge function.

so some how I want a secure table as well as cron job!

thanks for time!

please give your suggestions. 

I'm trying to solve this for long time.

**Comment by GaryAustin1** [comment]
Your description does not make much sense.
The function does not care what key you pass it to run.  Either anon or service role (or a user jwt) would meet the requirement to have a jwt in the authorization header for the edge function (if that is how you have it set).

Then your edge function sets the createClient as you desire to make calls to Supabase.

You could pass in another header with a secret_key you make up in the cron task and check for that in the edge function.  Or you could pass in the service_role key and then check for that in the authorization header to keep anon key from executing the function.

**Reply by Kgitman** [reply]
really sorry for bad description, and thanks for response @GaryAustin1
I followed your suggestions and make the changes below.

![image](https://github.com/user-attachments/assets/eee033d9-619f-4483-adb0-f74c69689524)

and in this SQL code I'm passing the service role key 

```
select
  cron.schedule(
    'invoke-function-every-minute',
    '* * * * *', -- every minute
    $$
    select
      net.http_post(
          url:='https://project-ref.supabase.co/functions/v1/function-name',
          headers:='{"Content-Type": "application/json", "Authorization": "Bearer eyj9quetihtehg.keueuijngeigeiieiuheihei6bfg56FHG.ehK8KH-VSjkej"}'::jsonb,
          body:=concat('{"time": "', now(), '"}')::jsonb
      ) as request_id;
    $$
  );
```


But I'm not getting the `service_role_key` in my `edge function`

for checking I wrote console log you can see in image 

`console.log('Authorization header:', authHeader);`

And In this above log output I got there is `anon_key` not `service_role_key`


`In short I want to secure my table as well as I want to run the cron-job if you or anyone have different approach to do this than please let me know! `

again thanks for your time!

**Reply by GaryAustin1** [reply]
Why do you need to pass the service role key versus just have a secret key header you check and then just use the service_role key from  env.  Or do you want to handle both anon and service_role calls to this edge function?

What is getAuthKey()

And in your cron task the thing you put in the Authorization header is the service_role key?

**Reply by Kgitman** [reply]
@GaryAustin1 thank you!

- I'm passing the service_role_key because I want to instruct my code that if you got the `service_role_key` than and than run the cron job otherwise not.

- If it not `service_role_key` than use the `anon_key` which denied the opration.

- I used the `getAuthKey()` for comparing the incomig `service_role_key`.

- yes in my cron task the thing I put in the Authorization header is the `service_role_key`.

**I don't know that my approaches are right or wrong but what I want to do is I'm describing below :**

**I just want that my tables are secure as well as my cron job should be also run. and my anon key is exposed in the REST APIs so I don't want to use anon_key for any other operation.** 

**If you have anything than please guide.**

thanks agian for your time.

**Reply by GaryAustin1** [reply]
You did not answer what getAuthKey() is or show code for it.

**Reply by Kgitman** [reply]
getAuthkey() is function which accepting the header from main edge function.

![image](https://github.com/user-attachments/assets/79b324e0-8fcf-463a-9b3a-72ff5680ee05)

**Reply by GaryAustin1** [reply]
I have no idea what a main function is in relation to your code.
![image](https://github.com/user-attachments/assets/b390ae03-785b-467e-bd17-d8a4425d6561)
You should just be able to compare like you want right after the red underline (code from the docs).

**Reply by Kgitman** [reply]
is it the cron job doc? can you share the link?

**Reply by GaryAustin1** [reply, marked as answer]
Your cron job if that is the service_role key seems fine.
https://supabase.com/docs/guides/functions/auth

**Reply by Kgitman** [reply]
thanks @GaryAustin1 

do you have any other solution in mind for the below problem?

> I just want that my tables are secure as well as my cron job should be also run. and my anon key is exposed in the REST APIs so I don't want to use anon_key for any other operation.

**Reply by GaryAustin1** [reply]
I already suggested passing a secret header from the cron task with a code it in.  Then check that header for the code in the function with rec.headers.get.
But what you want to do is fine, sending the service role key from cron in the authorization header and getting it with rec.headers.get.   You must be doing something wrong with the call to the header function to get the header or something else strange is going on.

**Reply by Kgitman** [reply]
Thank you so much @GaryAustin1 it's working now.

**Reply by GaryAustin1** [reply]
Can you say what you did to fix it?

**Reply by Kgitman** [reply]
From authorization I'm not extracting the the key differently and bearer differently that's why it's give me messege of violates the RLS, because it's not getting the KEY properly.

Below are the changes I made:

![image](https://github.com/user-attachments/assets/f2d407e9-f438-4fcc-ae79-fcf2b1dafe4a)
![image](https://github.com/user-attachments/assets/45004463-a962-47d8-8944-ff94cdce0db3)

**Reply by lorikku** [reply]
> I already suggested passing a secret header from the cron task with a code it in. Then check that header for the code in the function with rec.headers.get. But what you want to do is fine, sending the service role key from cron in the authorization header and getting it with rec.headers.get. You must be doing something wrong with the call to the header function to get the header or something else strange is going on.

I want to pass the service role key from cron, but it feels illegal just pasting it in there like this
```
SELECT cron.schedule(
  'test',
  '0 0 * * *',
  $$
  SELECT net.http_post(
    'FUNCTION_URL',
    '{}'::jsonb,
    '{
      "Content-Type": "application/json",
      "Authorization": "Bearer SERVICE_ROLE_KEY"
    }'::jsonb
  );
  $$
);
```

Is this safe? What if the service role key changes? Is there a way to access this as an env var when creating a new cron?

**Reply by GaryAustin1** [reply]
Most put the keys in Vault and the select from that in their function code.   You still have to change Vault but at least it is one clear spot.

There is currently no environment options for this data.

**Reply by lorikku** [reply]
So no way to use env variables in SQL query/cron, understood.
